packages/salesforcedx-vscode-core/src/context/workspaceContext.ts

```diff
@@ -0,0 +1,123 @@
+ * Manages the context of a workspace during a session with an open SFDX project.
+ */
+export class WorkspaceContext {
+  private subscribers = new Set<OrgSubscriber>();
+  private cliConfigWatcher?: vscode.FileSystemWatcher;
+  private sessionConnections = new Map<string, Connection>();
+  private _orgUsername?: string;
+  private _orgAlias?: string;
+
+  private static instance?: WorkspaceContext;
+
+  private constructor() {}
+
+  public static async initialize(context: vscode.ExtensionContext) {
+    const instance = new WorkspaceContext();
+    const cliConfigPath = join(getRootWorkspacePath(), SFDX_FOLDER, SFDX_CONFIG_FILE);
+    const bindedHandler = () => instance.handleCliConfigChange();
+
+    instance.cliConfigWatcher?.dispose();
+    instance.cliConfigWatcher = vscode.workspace.createFileSystemWatcher(cliConfigPath);
+    instance.cliConfigWatcher.onDidChange(bindedHandler);
+    instance.cliConfigWatcher.onDidCreate(bindedHandler);
+    instance.cliConfigWatcher.onDidDelete(bindedHandler);
+    context.subscriptions.push(instance.cliConfigWatcher);
+
+    await instance.handleCliConfigChange();
+
+    this.instance = instance;
+  }
+
+  public static get(): WorkspaceContext {
+    if (!this.instance) {
+      telemetryService.sendException(
+        'WorkspaceContextException',
+        'Workspace context has not been initialized'
+      );
+      throw new Error(nls.localize('error_workspace_context_init'));
+    }
+    return this.instance;
+  }
+
+  /**
+   * Subscribe to when the default org is set.
+   *
+   * @param subscriber Object to be notified
+   * @param notifyNow Notify the subscriber immediately after subscribing
+   */
+  public subscribe(subscriber: OrgSubscriber, notifyNow = false) {
+    this.subscribers.add(subscriber);
+    if (notifyNow) {
+      this.notifySubscriber(subscriber);
+    }
+  }
+
+  public async getConnection(): Promise<Connection> {
+    if (!this._orgUsername) {
+      throw new Error(nls.localize('error_no_default_username'));
+    }
+
+    let connection = this.sessionConnections.get(this._orgUsername);
+    if (!connection) {
+      connection = await Connection.create({
+        authInfo: await AuthInfo.create({ username: this._orgUsername })
+      });
+      this.sessionConnections.set(this._orgUsername, connection);
+    }
+
+    return connection;
+  }
+
+  private async handleCliConfigChange() {
+    const usernameOrAlias = await OrgAuthInfo.getDefaultUsernameOrAlias(false);
+
+    if (usernameOrAlias) {
+      this._orgUsername = await OrgAuthInfo.getUsername(usernameOrAlias);
+      this._orgAlias =
+        usernameOrAlias !== this._orgUsername ? usernameOrAlias : undefined;
+    } else {
+      this._orgUsername = undefined;
+      this._orgAlias = undefined;
+    }
+
+    setupWorkspaceOrgType(this._orgUsername)
+      .catch(e => console.error('Error setting VS Code org type context'));
+
+    for (const subscriber of this.subscribers) {
+      this.notifySubscriber(subscriber);
```

Comment:
It's your call, but we could use `vscode.EventEmitter` or node event emitter instead of a custom implementation. You'll get things like `dispose` for free.

Reply:
I briefly thought about using the vscode EventEmitter but wasn't sure if there was a major benefit to doing one way over the other. I'll take a closer look tomorrow to see what it'd be like.

What's the preference of using vscode.EventEmitter vs. NodeJS build-in?

Reply:
@brpowell https://code.visualstudio.com/api/references/vscode-api#EventEmitter%3CT%3E I haven't tried if it works across extensions, but the document says 
>Use this class if you want to provide event from within your extension, for instance inside a TextDocumentContentProvider or when providing API to other extensions.

also, `vscode.EventEmitter` is used across a lot of VS Code APIs, so it would be beneficial if we are providing it as an interface in our APIs for consistency.

Reply:
@xyc thanks for the suggestion, I went ahead and updated it to use the vscode.EventEmitter impl.